homeassistant/components/recorder/services.py

```diff
@@ -35,15 +36,28 @@
 ATTR_DOMAINS = "domains"
 ATTR_ENTITY_GLOBS = "entity_globs"
 
-SERVICE_PURGE_ENTITIES_SCHEMA = vol.Schema(
-    {
-        vol.Optional(ATTR_DOMAINS, default=[]): vol.All(cv.ensure_list, [cv.string]),
-        vol.Optional(ATTR_ENTITY_GLOBS, default=[]): vol.All(
-            cv.ensure_list, [cv.string]
+SERVICE_PURGE_ENTITIES_SCHEMA = vol.All(
+    vol.Schema(
+        {
+            vol.Optional(ATTR_ENTITY_ID, default=[]): cv.entity_ids,
+            vol.Optional(ATTR_DOMAINS, default=[]): vol.All(
+                cv.ensure_list, [cv.string]
+            ),
+            vol.Optional(ATTR_ENTITY_GLOBS, default=[]): vol.All(
+                cv.ensure_list, [cv.string]
+            ),
+            vol.Optional(ATTR_KEEP_DAYS, default=0): cv.positive_int,
+        }
+    ),
+    vol.Any(
+        vol.Schema({vol.Required(ATTR_ENTITY_ID): vol.IsTrue()}, extra=vol.ALLOW_EXTRA),
+        vol.Schema({vol.Required(ATTR_DOMAINS): vol.IsTrue()}, extra=vol.ALLOW_EXTRA),
+        vol.Schema(
+            {vol.Required(ATTR_ENTITY_GLOBS): vol.IsTrue()}, extra=vol.ALLOW_EXTRA
         ),
```

Comment:
I was hoping we had something off the shelf in config_validation for this, but I didn't find anything

Reply:
Yeah I didn't find either. I'm not that experienced with voluptuous so I'm not sure if this is the most elegant way to do this, but was the first thing I stumbled into that seemed to achieve what I was looking for.

Reply:
You could write your own validator like https://github.com/home-assistant/core/blob/1d6ae01baab69957e1bf7f26fc00564a4de9df9c/homeassistant/helpers/config_validation.py#L166 but I'm not sure its going to be any better as its likely just as many lines.